tapdb/sqlc/queries/transfers.sql

```diff
@@ -23,11 +23,36 @@ INSERT INTO asset_transfer_outputs (
     amount, serialized_witnesses, split_commitment_root_hash,
     split_commitment_root_value, proof_suffix, num_passive_assets,
     output_type, proof_courier_addr, asset_version, lock_time,
-    relative_lock_time
+    relative_lock_time, proof_deliver_complete
 ) VALUES (
-    $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
+    $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16
 );
 
+-- name: SetTransferOutputProofDeliveryStatus :exec
+WITH target(output_id) AS (
+    SELECT output_id
+    FROM asset_transfer_outputs output
+    JOIN asset_transfers
+      ON output.transfer_id = asset_transfers.id
+    JOIN chain_txns
+      ON asset_transfers.anchor_txn_id = chain_txns.txn_id
+    JOIN script_keys
+      ON output.script_key = script_keys.script_key_id
+    WHERE (chain_txns.txid = sqlc.narg('anchor_tx_id') OR sqlc.narg('anchor_tx_id') IS NULL)
+      AND (script_keys.tweaked_script_key = sqlc.narg('script_key') OR sqlc.narg('script_key') IS NULL)
+      AND (output.script_key_local = sqlc.narg('script_key_local') OR sqlc.narg('script_key_local') IS NULL)
+      AND (output.amount = sqlc.narg('amount') OR sqlc.narg('amount') IS NULL)
+      AND (output.serialized_witnesses = sqlc.narg('serialized_witnesses') OR sqlc.narg('serialized_witnesses') IS NULL)
```

Comment:
I don't think we should use any of these serialized values as query elements.
What I think we need is a way to uniquely address a specific transfer output. So we should add a new column `position` (trying to avoid the use of the word "index" to not make this easy to confuse with the output index of a transaction) on the `asset_transfer_outputs` table that for new transfer outputs corresponds to the index in the `OutboundParcel.Outputs` slice. And for existing transfer outputs we can just use the primary key to back fill, since we just need a value that's unique per transfer.

Then this whole query can just query on the `anchor_utxo` ID (outpoint instead of just TXID) and the `position`.

So basically this:
```sql
-- The output position indicates the order of the output in the list of outputs
-- for a given transfer. We'll be inserting an actual value in the next query,
-- so we just start with -1.
ALTER TABLE asset_transfer_outputs ADD COLUMN position INTEGER NOT NULL DEFAULT -1;

-- Update the position to be the same as the output id for existing entries.
-- We'll use the position integer as a uniquely identifiable number of an output
-- within a transfer, so setting the default to the output_id is just to make
-- sure we have a unique value that also satisfies the unique constraint we add
-- below.
UPDATE asset_transfer_outputs SET position = CAST(output_id AS INTEGER)
WHERE position = -1;

-- We need to be able to upsert outputs, so we need a unique constraint on
-- (output_id, position).
CREATE UNIQUE INDEX asset_transfer_outputs_output_id_position_unique
ON asset_transfer_outputs (
    output_id, position
);
```

Reply:
I'm not sure if the new unique constraint `asset_transfer_outputs_output_id_position_unique` is necessary. `output_id` is already a primary key, so its unique. And so (output_id, position) will be unique because `output_id` is a primary key.

Reply:
The deliver subroutine doesn't currently have access to the output index in `OutboundParcel.Outputs`. We use:

```
err := fn.ParSlice(ctx, pkg.OutboundPkg.Outputs, deliver)
```

Would be necessary to have a version of `fn.ParSlice` which passes in the index. So we can get the position to `p.cfg.ExportLog.ConfirmProofDelivery`. That's why I didn't go down the `position` route but rather specify everything. But I agree that ultimately a position index would be more solid. I will work on this.

Reply:
The unique index is more of a safety net to make sure we don't screw anything up, since we rely on the position being unique for each transfer.

> The deliver subroutine doesn't currently have access to the output index ...

Ah, yes, because of the `fn.ParSlice`. I think it makes sense to just add a new `Position` field on the `OutboundParcel.Output` struct, since it will be stored in the database as well. Then we can still make use of the async/unordered nature of `fn.ParSlice`

Reply:
> The unique index is more of a safety net to make sure we don't screw anything up, since we rely on the position being unique for each transfer.

Ah got it. A unique constraint on transaction id and position perhaps.

Reply:
Ah, yes, of course... Sorry, seems I was a bit confused here. `output_id` doesn't make any sense, should be `transfer_id, position`.